### PR TITLE
Flatten Avro Schema to make it optimal for ORC

### PR DESCRIPTION
**Pull request by abti**
@chavdar Please review

### CODE REVIEW

**Comment by kadaan** [comment]
@abti ORC has great support for complex types.  Why would you want to flatten the avro schema when converting the data to ORC?

**Comment by zliu41** [comment]
@kadaan flattening improves query response time for many types of queries by a noticeable margin

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroFlattener.class);
+
+  /***
+   * Flatten the Schema to un-nest recursive Records (to make it optimal for ORC)
+   * @param schema Avro Schema to flatten
+   * @return Flattened Avro Schema
+   */
+  public static Schema flatten(Schema schema) {
+    // Validate input
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: There are many unnecessary comments such as this one, line 43, 53, 156, etc. These are obvious from the code so please remove them

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    Preconditions.checkNotNull(schema);
+
+    // Flatten schema
+    Schema flattenedSchema = flatten(schema, 0);
+
+    LOG.info("Original Schema : " + schema);
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Please change this to debug. A schema can be thousands of characters and if every schema is printed it takes a lot of space.

**Comment by ibuenros** [comment]
@kadaan Hive does not support predicate pushdown for ORC with nested fields:
https://issues.apache.org/jira/browse/HIVE-7214

**Comment by zliu41** [comment]
@abti should this PR include a Converter?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+  private static List<Schema.Field> flattenField(Schema.Field f, LinkedList<String> lineage, int level,
+      boolean shouldWrapInOption) {
+    Preconditions.checkNotNull(f);
+    Preconditions.checkNotNull(f.schema());
+    Preconditions.checkNotNull(f.name());
+
+    List<Schema.Field> flattenedFields = new ArrayList<>();
+
+    // Set lineage
+    if (null == lineage) {
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Doesn't seem possible for lineage to be null.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+          flattenedFields.addAll(flattenField(field, lineage, level+1, false));
+        }
+        lineage.removeLast();
+      }
+    }
+    // If field.Type = OPTION, un-nest its fields and return them
+    else {
+      if (isOfOptionType(f.schema())) {
+        Schema first = f.schema().getTypes().get(0);
+        Schema second = f.schema().getTypes().get(1);
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: What if there are more than 2 options?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        if (null != schema.getTypes() && schema.getTypes().size() > 0) {
+          for (Schema oldUnionMember : schema.getTypes()) {
+            // Its member might still recursively contain records
+            flattenedUnionMembers.add(flatten(oldUnionMember, level + 1));
+          }
+        }
+        flattenedSchema = Schema.createUnion(flattenedUnionMembers);
+        break;
+      default:
+        String exceptionMessage = String.format("Schema flattening failed for \"%s\" ", schema);
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Can you just have `case`s for RECORD, UNION, MAP and ARRAY, and use `default` for all others?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    List<Schema.Field> flattenedFields = new ArrayList<>();
+
+    // Set lineage
+    if (null == lineage) {
+      lineage = new LinkedList<>();
+    }
+    lineage.add(f.name());
+
+    // If field.Type = RECORD, un-nest its fields and return them
+    if (Schema.Type.RECORD.equals(f.schema().getType())) {
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: This seems to be similar logic as the `case RECORD` in line 95. Can you put it in a method `flattenRecord`?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+    // Copy schema metadata
+    copyProperties(schema, flattenedSchema);
+
+    return flattenedSchema;
+  }
+
+  /***
+   * Flatten Record field, and compute a list of flattened fields
+   * @param f Field to flatten
+   * @param lineage Lineage represented as a List of Strings
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Can you explain what `lineage` is using an example. It's not very clear to me.

**Comment by abti** [comment]
@kadaan Like ibuenros mentioned, Hive does not support pushdown for ORC for nested fields 

@zliu41 We have phased the overall effort, Converter/Writer and Source will follow soon

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 3 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Yes, it won't be null

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 4 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: If there are two fields which are options, both get un-nested

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 5 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: We want default to throw exception. Modeled it around an Avro's internal class

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 7 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: I will add an example

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 1 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: I will remove them, though I like to have a few extra comments :)

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 2 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: I will change this.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 6 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Moved Record and Union flattening to methods

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+public class AvroFlattener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroFlattener.class);
+
+  /***
+   * Flatten the Schema to un-nest recursive Records (to make it optimal for ORC)
+   * @param schema Avro Schema to flatten
+   * @return Flattened Avro Schema
+   */
+  public static Schema flatten(Schema schema) {
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Please make the methods instance-level. This will allow us to easily add configs and similar in the future.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.commons.lang3.StringUtils;
+import org.codehaus.jackson.JsonNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+public class AvroFlattener {
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Please add a high-level description of how the flattening works, what gets flattened, naming, metadata, etc.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        break;
+      case NULL:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case RECORD:
+        List<Schema.Field> flattenedFields = new ArrayList<>();
+        if (null != schema.getFields() && schema.getFields().size() > 0) {
+          // Recursively flatten fields
+          for (Schema.Field oldField : schema.getFields()) {
+            LinkedList<String> lineage = new LinkedList<>();
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Shouldn't the lineage be a list of Schemas rather than just names. That give more flexibility, e.g. if you want to concatenate doc strings.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        break;
+      case UNION:
+        List<Schema> flattenedUnionMembers = new ArrayList<>();
+        if (null != schema.getTypes() && schema.getTypes().size() > 0) {
+          for (Schema oldUnionMember : schema.getTypes()) {
+            // Its member might still recursively contain records
+            flattenedUnionMembers.add(flatten(oldUnionMember, level + 1));
+          }
+        }
+        flattenedSchema = Schema.createUnion(flattenedUnionMembers);
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: 1. Union of a [record_type, null] and [null, record_type] should be handled as a special case. It should be flattened to the result of the flattening of record_type and all fields should be optional.
2. [a lower priority, might add as an extension in the future] The case of union of [record_type_1, record_type_2, ...] may be handled by combining the flattened record_type_1, record_type_2, ... and using pre-pending the name of the record types to each field name for disambiguation.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    return flattenedFields;
+  }
+
+  /***
+   * Check if the Avro Schema is of type OPTION
+   * ie. [null, RECORD] or [RECORD, null]
+   * @param schema Avro Schema to check
+   * @return true if Avro Schema is of type OPTION else false
+   */
+  private static boolean isOfOptionType(Schema schema) {
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: nit: it might be worthwhile to change that the method returns the record type if it is optional type as it is needed anyway.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+        lineage.removeLast();
+      }
+    }
+    // If field.Type = OPTION, un-nest its fields and return them
+    else {
+      if (isOfOptionType(f.schema())) {
+        Schema first = f.schema().getTypes().get(0);
+        Schema second = f.schema().getTypes().get(1);
+        Schema record = first;
+        boolean wrapInOption = Schema.Type.NULL.equals(first.getType());
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: wrapInOption should be true in the case of [record_type, null] too.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        if (StringUtils.isNotBlank(flattenSource)) {
+          field.addProp("flatten_source", flattenSource);
+        }
+        for (Map.Entry<String, JsonNode> entry : f.getJsonProps().entrySet()) {
+          field.addProp(entry.getKey(), entry.getValue());
+        }
+        flattenedFields.add(field);
+      }
+    }
+    lineage.removeLast();
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Modifying "lineage" and trying to revert it to its original value is very error-prone. Someone will later modify the code, forget to do it and all kinds of bad things may happen. I think it is best to make "lineage" immutable and make copies if necessary to modify it.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+   *            "name" : "nestedFieldString",
+   *            "type" : "string"
+   *          }, {
+   *            "name" : "nestedFieldInt",
+   *            "type" : "int"
+   *          } ]
+   *       }
+   *     }]
+   * }
+   * The expected output schema is:
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: The output schema doesn't have "nestedRecordName". Is this expected?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 9 above)
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Also is there any restrictions, e.g., can it handle recursive schema?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+
+  private static Schema flatten(Schema schema, boolean shouldPopulateLineage) {
+    Schema flattenedSchema;
+
+    // Process all Schema Types
+    // (Primitives are simply cloned)
+    switch (schema.getType()) {
+      case ARRAY:
+        // Array might be an array of recursive Records, flatten them
+        flattenedSchema = Schema.createArray(flatten(schema.getElementType(), false));
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Can you explain in the comments why use `false` here. Same for `MAP`.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+   * @return Flattened Record Schema
+   */
+  private static Schema flattenRecord(Schema schema, boolean shouldPopulateLineage) {
+    Preconditions.checkNotNull(schema);
+    Preconditions.checkArgument(Schema.Type.RECORD.equals(schema.getType()));
+
+    Schema flattenedSchema;
+
+    List<Schema.Field> flattenedFields = new ArrayList<>();
+    if (null != schema.getFields() && schema.getFields().size() > 0) {
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: This line is unnecessary since `schema.getFields()` won't return null. Same for line 132.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+   *     }]
+   * }
+   * The expected output schema is:
+   * {
+   *    "type" : "record",
+   *    "name" : "parentRecordName",
+   *    "fields" : [ {
+   *      "name" : "parentFieldRecord__nestedFieldString",
+   *      "type" : "string",
+   *      "flatten_source" : "parentFieldRecord.nestedFieldString"
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Seems `name` already contains all information in `flatten_source`. Why is `flatten_source` needed?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+      }
+    }
+    lineage.removeLast();
+
+    return flattenedFields;
+  }
+
+  /***
+   * Check if the Avro Schema is of type OPTION
+   * ie. [null, RECORD] or [RECORD, null]
```
Comment (zliu41) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Is it possible to have a schema of type [RECORD, String, null]?

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 8 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Changed it

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 9 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Added

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 12 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Changed to Optional<Schema> and made use of it for checking / getting Record

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 13 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Changed

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 14 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Made it immutable

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 15 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Yes, thats expected. Earlier we were adding that to output but after a discussion we removed it.

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 16 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: I will add some comments, but in quick summary: shouldPopulateLineage is set to true if the field is going to be flattened and moved up the hierarchy - so that lineage information can be tagged to it. Which happens when there is Record within Record OR Record within Option with Record and so on, however not when there is Record within Map within Record

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 17 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Removed

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 18 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: It came in as a requirement

File: gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java (hunk 19 above)
Comment (abti) [inline gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java]: Yes, but we do not consider it as an option

File: gobblin-utility/src/test/java/gobblin/util/AvroFlattenerTest.java [hunk 20]
```diff
@@ -0,0 +1,303 @@
+package gobblin.util;
+
+import org.apache.avro.Schema;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+
+public class AvroFlattenerTest {
```
Comment (chavdar) [inline gobblin-utility/src/test/java/gobblin/util/AvroFlattenerTest.java]: I think we should have some additional tests:
* Record within Record within Record
* Negative cases involving arrays or records

I think it will be easier to write tests if instead of inlining the schemas as string (lots of " escaping), schemas are loaded from resources: testX() reads the resources X.input.json and X.expected.json and verifies that flatten(X.input) == X.expected. Then we can easily add more tests.

Comment (abti) [inline gobblin-utility/src/test/java/gobblin/util/AvroFlattenerTest.java]: Moved json inputs to resources, and added more test cases (positive and negative)

**Comment by chavdar** [comment]
One minor comment. Other than that, LGTM.

**Comment by abti** [comment]
Thanks for the reviews everyone. Also addressed the minor comment by Chavdar. Merging
